# Genetic Analysis of a Metazoan Pathway using Transcriptomic Phenotypes
#### Authors: David Angeles-Albores, Carmie Puckett-Robinson, Brian A. Williams, Barbara Wold and Paul W. Sternberg

## Abstract
RNA-seq is commonly used to identify genetic modules that respond to a perturbation. Although transcriptomes have been mainly used for target gene discovery, their quantitative nature makes them attractive structures with which to study genetic interactions. To understand whether whole-organism RNA-seq is suitable for genetic pathway reconstruction, we sequenced the transcriptome of four single mutants and two double mutants of the hypoxia pathway in *C. elegans*. By comparing the expression levels of double mutants with their corresponding single mutants, we were able to determine, on a genome-wide level, that EGL-9 acts along VHL-1-dependent and independent branches to inhibit HIF-1. We were also able to observe transcriptome-wide suppression of the *egl-9(lf)* phenotype in an *egl-9(lf) hif-1(lf)* double mutant. As a by-product of our analysis, we identified a core hypoxic response consisting of 355 genes, and 45 genes that have *hif-1*-independent, *vhl-1*-dependent expression. Finally, we are able to identify 31 genes that exhibit non-canonical epistasis: for these genes, 24 *vhl-1(lf)* mutants show opposing effects to *egl-9(lf)* mutants, but the *egl-9(lf);vhl-1(lf)* exhibits the *egl-9(lf)* phenotype. We suggest that this non-canonical epistasis reflects unexplored aspects of the hypoxia pathway. We discuss the implications, benefits and advantages of using transcriptomic phenotypes to perform pathway analysis.

# File Structure
```
mprsq
│   README.md
│   kallisto_commands_sh    
│
└───input
└───experimental_docs
└───sleuth_all_adjusted
└───src
└───output
└───docs
```


## Contact
pws@caltech.edu

## Technical assistance
dangeles@caltech.edu

## Acknowledgements
