{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "button": false,
    "collapsed": false,
    "deletable": true,
    "new_sheet": false,
    "run_control": {
     "read_only": false
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of na genes: 1005\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import morgan as morgan\n",
    "import argparse\n",
    "import numpy as np\n",
    "\n",
    "# get the read number from terminal\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"-n\", \"--number\", type=int,\n",
    "                    help=\"Number of files that were processed in this sample\")\n",
    "args = parser.parse_args()\n",
    "if args.number:\n",
    "    number = args.number\n",
    "else:\n",
    "    number=-1\n",
    "\n",
    "\n",
    "# load parameters:\n",
    "single_mutants = ['b', 'c', 'd', 'e', 'g']\n",
    "double_mutants = {'a' : 'bd', 'f':'bc'}\n",
    "\n",
    "# initialize thomas object and load all params\n",
    "thomas = morgan.hunt('target_id', 'b', 'tpm', 'qval')\n",
    "thomas.add_genmap('../input/library_genotype_mapping.txt', comment='#')\n",
    "thomas.add_single_mutant(single_mutants)\n",
    "thomas.add_double_mutants(['a', 'f'], ['bd', 'bc'])\n",
    "thomas.set_qval()\n",
    "\n",
    "# Add the tpm files: \n",
    "kallisto_loc = '../input/kallisto_all/'\n",
    "thomas.add_tpm(kallisto_loc, '/kallisto/abundance.tsv', '')\n",
    "\n",
    "# Make all possible combinations of WT, X\n",
    "combs = {}\n",
    "for gene in thomas.genmap.genotype.unique():\n",
    "    if gene != 'wt':\n",
    "        combs[gene] = 'WT_'+gene+'/'\n",
    "\n",
    "# load all the beta values for each genotype:\n",
    "sleuth_loc = '../sleuth/'\n",
    "thomas.add_betas(sleuth_loc, 'betas.csv', combs)\n",
    "thomas.filter_data(0, 0.1)\n",
    "\n",
    "# extract relevant information\n",
    "mat = np.matrix([['all', thomas.beta['a'].ens_gene.unique().shape[0], number, 'total_genes'],\n",
    "                ['all', thomas.beta_filtered['a'].ens_gene.unique().shape[0], number, 'filtered_genes']])\n",
    "\n",
    "for key in thomas.beta_filtered.keys():\n",
    "    df = thomas.beta_filtered[key]\n",
    "    sig = df[df.qval < 0.1].ens_gene.unique().shape[0]\n",
    "    row1 = [key, df.ens_gene.unique().shape[0], number, 'total_genes']\n",
    "    row2 = [key, sig, number, 'significant_genes']\n",
    "    mat = np.vstack((mat, row1))\n",
    "    mat = np.vstack((mat, row2))\n",
    "\n",
    "df = pd.DataFrame(mat, columns=['pair','a_value','analysis','reads'])\n",
    "\n",
    "# perform analyses\n",
    "fast = morgan.brenner('single_mutant_analysis', thomas)\n",
    "bayes = morgan.mcclintock('single_mutant_analysis', thomas, progress=False)\n",
    "\n",
    "# super melt function\n",
    "def supermelt(df, analysis, reads, col='test_value'):\n",
    "    \"\"\"A function to melt a dataframe into a predetermined format.\"\"\"\n",
    "    new = pd.melt(df, id_vars='corr_with', var_name='genotype', value_name=col)\n",
    "    new = new[new[col].abs() > 0]\n",
    "    new['pair'] = new.genotype + new.corr_with\n",
    "    new = new[new.genotype != new.corr_with]\n",
    "    new['analysis'] = analysis\n",
    "    new['reads'] = reads\n",
    "    return new[['pair', col, 'analysis', 'reads']]\n",
    "\n",
    "# melt everything\n",
    "rho = supermelt(fast.rho, 'spearman', number, 'a_value')\n",
    "hyper_plus = supermelt(fast.hyper_plus, 'hyper_plus', number, 'a_value')\n",
    "hyper_minus = supermelt(fast.hyper_minus, 'hyper_minus', number, 'a_value')\n",
    "bayes_primary = supermelt(bayes.robust_slope, 'bayes_primary', number, 'a_value')\n",
    "\n",
    "bayes.secondary_slope['corr_with'] = thomas.single_mutants\n",
    "bayes_secondary = supermelt(bayes.secondary_slope, 'bayes_secondary', number, 'a_value')\n",
    "\n",
    "# save it to a spreadsheet\n",
    "if os.path.isfile('../input/downsampling.csv') == False:\n",
    "    rho.to_csv('../input/downsampling.csv', index=False)\n",
    "    hyper_plus.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    hyper_minus.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    bayes_primary.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    bayes_secondary.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    df.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "else:\n",
    "    rho.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    hyper_plus.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    hyper_minus.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    bayes_primary.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    bayes_secondary.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)\n",
    "    df.to_csv('../input/downsampling.csv', mode='a', index=False, header=False)    "
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
